e2e/htmlSnapshot.ts

```diff
@@ -0,0 +1,188 @@
+import type {Locator} from '@playwright/test';
+
+export type HTMLAttribute = {name: string; value: string};
+export type HTMLNode =
+	| null
+	| string
+	| {
+			tagName: string;
+			childNodes: HTMLNode[];
+			attributes: HTMLAttribute[];
+	  };
+
+export const htmlStructure = (locator: Locator): Promise<HTMLNode> =>
+	locator.evaluate((element) => {
+		const isElement = (node: Node): node is Element => node.nodeType === node.ELEMENT_NODE;
+		const isInput = (node: Element): node is HTMLInputElement => node.tagName === 'INPUT';
+		const isTextArea = (node: Element): node is HTMLTextAreaElement => node.tagName === 'TEXTAREA';
+		const recFn = (node: Node) => {
+			if (!isElement(node)) {
+				if (node.nodeType === node.TEXT_NODE) {
+					return node.nodeValue;
+				}
+				// Note: ignore comments
+				return null;
+			}
+			const tagName = node.tagName.toLowerCase();
+			const textarea = isTextArea(node);
```

Comment:
minor: could be renamed to ```isNodeArea``` or is ```isNodeTextArea``` to respect naming conventions on boolean variables